File: code/Medium/Stack/Evaluate Reverse Polish Notation.js
// https://leetcode.com/problems/evaluate-reverse-polish-notation/
/**
 * @param {string[]} tokens
 * @return {number}
 */
var evalRPN = function (tokens) {
  let arr = [];

  for (let i = 0; i < tokens.length; i++) {
    if (tokens[i] == "+") {
      let a = arr[arr.length - 1];
      arr.pop();
      let b = arr[arr.length - 1];
      arr.pop();

      arr.push(b + a);
    } else if (tokens[i] == "-") {
      let a = arr[arr.length - 1];
      arr.pop();
      let b = arr[arr.length - 1];
      arr.pop();
      arr.push(b - a);
    } else if (tokens[i] == "/") {
      let a = arr[arr.length - 1];
      arr.pop();
      let b = arr[arr.length - 1];
      arr.pop();

      arr.push(parseInt(b / a));
    } else if (tokens[i] == "*") {
      let a = arr[arr.length - 1];
      arr.pop();
      let b = arr[arr.length - 1];
      arr.pop();

      arr.push(b * a);
    } else {
      arr.push(tokens[i] * 1);
    }
  }
  return arr[0];
};

File: code/Medium/Stack/Evaluate Reverse Polish Notation.cpp
// https://leetcode.com/problems/evaluate-reverse-polish-notation/

class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        stack<int> st;

        for (int i = 0; i < tokens.size(); i++)
        {

            if (tokens[i] == "+")
            {
                int a = (int)st.top();
                st.pop();
                int b = (int)st.top();
                st.pop();

                st.push(a + b);
            }
            else if (tokens[i] == "-")
            {
                int a = (int)st.top();
                st.pop();
                int b = (int)st.top();
                st.pop();

                st.push(a - b);
            }
            else if (tokens[i] == "/")
            {
                int a = (int)st.top();
                st.pop();
                int b = (int)st.top();
                st.pop();

                st.push(a / b);
            }
            else if (tokens[i] == "*")
            {
                int a = (int)st.top();
                st.pop();
                int b = (int)st.top();
                st.pop();

                st.push(a * b);
            }
            else
            {
                st.push[(int)tokens[i]];
            }
        }

        return st.top();
    }
};